warehouse/accounts/views.py

```diff
@@ -816,6 +822,174 @@ def _get_two_factor_data(request, _redirect_to="/"):
     return two_factor_data
 
 
+@view_config(
+    route_name="accounts.verify-organization-role",
+    renderer="accounts/organization-invite-confirmation.html",
+    require_methods=False,
+    uses_session=True,
+    permission="manage:user",
+    has_translations=True,
+)
+def verify_organization_role(request):
+    token_service = request.find_service(ITokenService, name="email")
+    organization_service = request.find_service(IOrganizationService, context=None)
+    user_service = request.find_service(IUserService, context=None)
+
+    def _error(message):
+        request.session.flash(message, queue="error")
+        return HTTPSeeOther(request.route_path("manage.organizations"))
+
+    try:
+        token = request.params.get("token")
+        data = token_service.loads(token)
+    except TokenExpired:
+        return _error(request._("Expired token: request a new organization invite"))
+    except TokenInvalid:
+        return _error(request._("Invalid token: request a new organization invite"))
+    except TokenMissing:
+        return _error(request._("Invalid token: no token supplied"))
+
+    # Check whether this token is being used correctly
+    if data.get("action") != "email-organization-role-verify":
+        return _error(request._("Invalid token: not a collaboration invitation token"))
+
+    user = user_service.get_user(data.get("user_id"))
+    if user != request.user:
+        return _error(request._("Organization invitation is not valid."))
+
+    organization = organization_service.get_organization(data.get("organization_id"))
+    desired_role = data.get("desired_role")
+
+    organization_invite = organization_service.get_organization_invite_by_user(
+        organization.id, user.id
+    )
+    if not organization_invite:
+        return _error(request._("Organization invitation no longer exists."))
+
+    # Use the renderer to bring up a confirmation page
+    # before adding as contributor
+    if request.method == "GET":
+        return {
+            "organization_name": organization.name,
+            "desired_role": desired_role,
+        }
+    elif request.method == "POST" and "decline" in request.POST:
+        organization_service.delete_organization_invite(organization_invite.id)
+        submitter_user = user_service.get_user(data.get("submitter_id"))
+        organization.record_event(
+            tag="organization:organization_role:declined",
+            ip_address=request.remote_addr,
```

Comment:
In this instance we may want to either anonymize the ip_address on the event for the OrganizationEvent by hard coding to 0.0.0.0/127.0.0.1 or somehow be *sure* we don't leak it on display to Organization admins.

This is a potential vector for abuse/leak. I can just invite a bunch of people to an org and get their IPs if they decline (or accept I suppose!)

Reply:
@di @dstufft thoughts?